crates/project_model/src/cargo_workspace.rs

```diff
@@ -63,8 +63,10 @@ pub struct CargoConfig {
     /// when debugging isolated issues.
     pub no_sysroot: bool,
 
-    /// rustc private crate source
-    pub rustc_source: Option<AbsPathBuf>,
+    /// Text-form path to the rustc's private crates source, or "discover" to find them
+    /// automatically.
+    /// Keep the root path around so we can reference relative paths to the private crates source.
+    pub rustc_source: Option<(AbsPathBuf, String)>,
```

Comment:
Year, this feels pretty awkward, especially because this puts `String -> PathBuf` conversion at the wrong layer. It should happen when we read config, not when we try to use the data. Let's introduce a proper enum here

```
enum RustcSource {
    Path(AbsPathBuf),
    Discover,
}
```